tools/spy-detector/src/main.rs

```diff
@@ -0,0 +1,262 @@
+                "Tx" => {
+                    if msg_type == 0 {
+                        stats
+                            .inv_tx_received
+                            .fetch_add(1, atomic::Ordering::Relaxed);
+                    } else {
+                        stats.inv_tx_sent.fetch_add(1, atomic::Ordering::Relaxed);
+                    }
+                }
+                "WTx" => {
+                    if msg_type == 0 {
+                        stats
+                            .inv_wtx_received
+                            .fetch_add(1, atomic::Ordering::Relaxed);
+                    } else {
+                        stats.inv_wtx_sent.fetch_add(1, atomic::Ordering::Relaxed);
+                    }
+                }
+                "WitnessTx" => {
+                    if msg_type == 0 {
+                        stats
+                            .inv_witnesstx_received
+                            .fetch_add(1, atomic::Ordering::Relaxed);
+                    } else {
+                        stats
+                            .inv_witnesstx_sent
+                            .fetch_add(1, atomic::Ordering::Relaxed);
+                    }
+                }
+                _ => {} // Ignore other types
+            }
+        }
+    }
+}
+
+fn process_getdata_msg(
+    peer_map: &PeerMap,
+    msg: &net_msg::message::Msg,
+    msg_type: u32,
+    peer_id: u64,
+) {
+    let stats = peer_map
+        .entry(peer_id.to_string())
+        .or_insert_with(PeerStats::default);
+
+    if let net_msg::message::Msg::Inv(inv) = msg {
+        for inv_item in &inv.items {
+            match inv_item.inv_type() {
+                "WitnessTx" => {
+                    if msg_type == 0 {
+                        stats
+                            .getdata_witnesstx_received
+                            .fetch_add(1, atomic::Ordering::Relaxed);
+                    } else {
+                        stats
+                            .getdata_witnesstx_sent
+                            .fetch_add(1, atomic::Ordering::Relaxed);
+                    }
+                }
+                _ => {} // Ignore other types
+            }
+        }
+    }
+}
+
+fn process_tx_msg(peer_map: &PeerMap, msg_type: u32, peer_id: u64) {
+    let stats = peer_map
+        .entry(peer_id.to_string())
+        .or_insert_with(PeerStats::default);
+
+    if msg_type == 0 {
+        stats.tx_received.fetch_add(1, atomic::Ordering::Relaxed);
+    } else {
+        stats.tx_sent.fetch_add(1, atomic::Ordering::Relaxed);
+    }
+}
+
+fn process_connection_event(peer_map: &PeerMap, event: &str) {
+    if event.starts_with("closed") {
+        let peer_id = event.split(' ').nth(1).unwrap_or("");
+        if let Some(stats) = peer_map.remove(peer_id) {
+            println!("Connection closed for peer: {}", peer_id);
+            print_peer_stats(peer_id, &stats.1);
+        }
+    }
+}
+
+fn print_peer_stats(peer_addr: &str, stats: &PeerStats) {
```

Comment:
Instead of having a print_peer_stats functions, it might be better to implement `std::fmt::Display` for `PeerStats`.